src/solvers/flattening/arrays.cpp

```diff
@@ -388,52 +385,42 @@ void arrayst::update_index_map(bool update_all)
   }
   else
   {
-    for(std::set<std::size_t>::const_iterator
-        it=update_indices.begin();
-        it!=update_indices.end(); it++)
-      update_index_map(*it);
+    for(const auto & it : update_indices)
+      update_index_map(it);
+
     update_indices.clear();
   }
 
 #ifdef DEBUG
   // print index sets
-  for(index_mapt::const_iterator
-      i1=index_map.begin();
-      i1!=index_map.end();
-      i1++)
-    for(index_sett::const_iterator
-        i2=i1->second.begin();
-        i2!=i1->second.end();
-        i2++)
-      std::cout << "Index set (" << i1->first << " = "
-                << arrays.find_number(i1->first) << " = "
-                << from_expr(ns, "", arrays[arrays.find_number(i1->first)])
+  for(const auto & i1 : index_map)
+    for(const auto & i2 : i1.second)
+      std::cout << "Index set (" << i1.first << " = "
+                << arrays.find_number(i1.first) << " = "
+                << from_expr(ns, "", arrays[arrays.find_number(i1.first)])
                 << "): "
                 << from_expr(ns, "", *i2) << '\n';
   std::cout << "-----\n";
 #endif
 }
 
-void arrayst::add_array_constraints(
+void arrayst::add_array_constraints_equality(
   const index_sett &index_set,
   const array_equalityt &array_equality)
 {
   // add constraints x=y => x[i]=y[i]
 
-  for(index_sett::const_iterator
-      it=index_set.begin();
-      it!=index_set.end();
-      it++)
+  for(const auto & it : index_set)
   {
     index_exprt index_expr1;
```

Comment:
`index_exprt index_expr1(array_equality.f1, it);` instead of the multi-step assignment.